src/McBopomofo.cpp

```diff
@@ -374,12 +390,23 @@ void McBopomofoEngine::keyEvent(const fcitx::InputMethodEntry&,
       enterNewState(context, std::make_unique<InputStates::Empty>());
       context->updateUserInterface(fcitx::UserInterfaceComponent::InputPanel);
       context->updatePreedit();
+
       return;
     }
 
-    handleCandidateKeyEvent(context, key, maybeCandidateList);
-    context->updateUserInterface(fcitx::UserInterfaceComponent::InputPanel);
-    context->updatePreedit();
+    handleCandidateKeyEvent(
+        context, key, maybeCandidateList,
+        [this, context](std::unique_ptr<InputState> next) {
+          enterNewState(context, std::move(next));
+        },
+        []() {
+          // TODO(unassigned): beep?
+        });
+    if (dynamic_cast<InputStates::ChoosingCandidate*>(state_.get()) !=
+        nullptr) {
+      context->updateUserInterface(fcitx::UserInterfaceComponent::InputPanel);
```

Comment:
(No action required.) I realized that some code paths `handleCandidateKeyEvent` already calls `updateUserInterface` before your change, and it doesn't do that consistently, and it doesn't call `updatePreedit` either. It seems the operation is idempotent effectively, so this is fine.